### PR TITLE
Bus filters

### PR DESCRIPTION
**Pull request by dmdashenkov**
In this PR we move the concept of `Filter` from the level of `CommandBus` to the level of `Bus`.

This makes the flow of `Bus.post()` more fluent and general.

### CODE REVIEW

**Comment by dmdashenkov** [comment]
@armiol, PTAL.

File: server/src/main/java/io/spine/server/bus/DeadMessageHandler.java [hunk 1]
```diff
@@ -0,0 +1,44 @@
...
+
+/**
+ * A functional interface for the function which handles a dead message.
+ *
+ * <p>If a message with no target dispatchers found is passed to the bus, it will result in a call
+ * to this function. The function handles such message and produces an {@link Error} as a response.
+ *
+ * @author Dmytro Dashenkov
+ */
+public interface DeadMessageHandler<E extends MessageEnvelope<?>> {
```
Comment (armiol) [inline server/src/main/java/io/spine/server/bus/DeadMessageHandler.java]: Looks `@Internal`.

Comment (armiol) [inline server/src/main/java/io/spine/server/bus/DeadMessageHandler.java]: Also, if that's a functional interface, why isn't it a `Function<E, Error>`? Let's discuss.

File: server/src/main/java/io/spine/server/bus/DeadMessageHandler.java [hunk 2]
```diff
@@ -0,0 +1,44 @@
...
+
+    /**
+     * Handles the dead message in a bus-specific way and produces an {@link Error} for notifying
+     * the poster about the absence of dispatchers.
+     *
+     * @param message the dead message
+     * @return the {@link Error} describing the dead message
```
Comment (armiol) [inline server/src/main/java/io/spine/server/bus/DeadMessageHandler.java]: `the {@link Error} describing the dead message` I don't think that's what it is. Let's discuss.

File: server/src/main/java/io/spine/server/bus/Bus.java [hunk 3]
```diff
@@ -272,6 +329,57 @@ private void doPost(Iterable<E> envelopes, StreamObserver<IsSent> observer) {
         return messageConverter;
     }
 
+    public abstract static class AbstractBuilder<E extends MessageEnvelope<T>,
+            T extends Message,
+            B extends AbstractBuilder<E, T, B>> {
+
+        private final Queue<BusFilter<E>> filters;
```
Comment (armiol) [inline server/src/main/java/io/spine/server/bus/Bus.java]: You have `Queue` here and `Deque` in the `FilterChain` below. Let's have it unified.

File: server/src/main/java/io/spine/server/bus/FilterChain.java [hunk 4]
```diff
@@ -0,0 +1,83 @@
...
+ *
+ * <p>The {@link #close() close()} method closes all the underlying filters.
+ *
+ * @author Dmytro Dashenkov
+ */
+final class FilterChain<E extends MessageEnvelope<T>, T> implements BusFilter<E> {
+
+    private final Deque<BusFilter<E>> chain;
+
+    private boolean closed;
```
Comment (armiol) [inline server/src/main/java/io/spine/server/bus/FilterChain.java]: Looks `volatile`.

File: server/src/main/java/io/spine/server/bus/ValidatingFilter.java [hunk 5]
```diff
@@ -0,0 +1,61 @@
...
+ */
+final class ValidatingFilter<E extends MessageEnvelope<T>, T extends Message>
+        extends AbstractBusFilter<E> {
+
+    private final Bus<T, E, ?, ?> bus;
+    private final EnvelopeValidator<E> validator;
+
+    ValidatingFilter(Bus<T, E, ?, ?> bus) {
+        super();
+        this.bus = bus;
```
Comment (armiol) [inline server/src/main/java/io/spine/server/bus/ValidatingFilter.java]: Let's have an internal `IdConverter` in `Bus` and store the reference to it, not to the whole `Bus`.

File: server/src/main/java/io/spine/server/commandbus/CommandBus.java [hunk 6]
```diff
@@ -67,7 +70,7 @@
 
     private final CommandStore commandStore;
 
-    private CommandBusFilter filterChain;
+    private final Deque<BusFilter<CommandEnvelope>> filters;
```
Comment (armiol) [inline server/src/main/java/io/spine/server/commandbus/CommandBus.java]: This should be `FilterChain`.

File: server/src/main/java/io/spine/server/event/EventValidator.java [hunk 7]
```diff
@@ -0,0 +1,72 @@
...
+final class EventValidator implements EnvelopeValidator<EventEnvelope> {
+
+    private final MessageValidator messageValidator;
+
+    EventValidator(MessageValidator messageValidator) {
+        this.messageValidator = messageValidator;
+    }
+
+    @Override
+    public Optional<Error> validate(EventEnvelope envelope) {
```
Comment (armiol) [inline server/src/main/java/io/spine/server/event/EventValidator.java]: Please check `EventException` and its descendants.

File: server/src/main/java/io/spine/server/bus/EnvelopeValidator.java [hunk 8]
```diff
@@ -0,0 +1,43 @@
...
+ */
+
+package io.spine.server.bus;
+
+import com.google.common.base.Optional;
+import io.spine.base.Error;
+import io.spine.core.MessageEnvelope;
+
+/**
+ * A functional interface defining the validator for a {@link MessageEnvelope}.
```
Comment (armiol) [inline server/src/main/java/io/spine/server/bus/EnvelopeValidator.java]: Same as above. If that's a functional interface, let's turn it into one.

File: server/src/main/java/io/spine/server/event/EventValidator.java [hunk 9]
```diff
@@ -0,0 +1,72 @@
...
+        checkNotNull(envelope);
+
+        final Event event = envelope.getOuterObject();
+        Throwable result = null;
+        final List<ConstraintViolation> violations = messageValidator.validate(event);
+        if (!violations.isEmpty()) {
+            final InvalidEventException invalidEvent = onConstraintViolations(event, violations);
```
Comment (armiol) [inline server/src/main/java/io/spine/server/event/EventValidator.java]: Some lines are reported to be uncovered.

File: server/src/main/java/io/spine/server/failure/UnsupportedFailureException.java [hunk 10]
```diff
@@ -0,0 +1,51 @@
...
+import io.spine.type.TypeName;
+
+import static java.lang.String.format;
+
+/**
+ * Exception that is thrown when unhandled failure is thrown.
+ *
+ * @author Dmytro Dashenkov
+ */
+public class UnsupportedFailureException extends RuntimeException {
```
Comment (armiol) [inline server/src/main/java/io/spine/server/failure/UnsupportedFailureException.java]: That's weird. I don't think we have unsupported failures at all.

File: server/src/main/proto/spine/server/integration/grpc/integration_event_subscriber.proto [hunk 11]
```diff
@@ -21,23 +21,22 @@ syntax = "proto3";
...
 option java_generate_equals_and_hash = true;
 
-import "spine/core/response.proto";
+import "spine/options.proto";
+import "spine/core/bus.proto";
 import "spine/server/integration/integration_event.proto";
 
 // A service for notifying integration event subscribers.
-// A Bounded Context is a typical integration event subscriber.
+// A Bounded Context is a typical integration event subscriber.is
```
Comment (armiol) [inline server/src/main/proto/spine/server/integration/grpc/integration_event_subscriber.proto]: Please fix a typo.

File: server/src/main/java/io/spine/server/bus/Bus.java [hunk 12]
```diff
@@ -61,6 +63,15 @@
     private DispatcherRegistry<C, D> registry;
 
     /**
+     * The chain of filters for this bus.
+     *
+     * <p>This field is effectively final, but is initialized lazily.
+     *
+     * @see #filterChain() for the non-null filter chain value
+     */
+    private FilterChain<E, T> filterChain;
```
Comment (dmdashenkov) [inline server/src/main/java/io/spine/server/bus/Bus.java]: Theoretically, `E` may be or be not bounded by `T`. To ensure we have the same message type, not `MessageEnvelope<?>`, we introduce the second type parameter here.

File: server/src/main/java/io/spine/server/bus/ValidatingFilter.java (hunk 5 above)
Comment (dmdashenkov) [inline server/src/main/java/io/spine/server/bus/ValidatingFilter.java]: Done.

File: server/src/main/java/io/spine/server/event/EventValidator.java (hunk 9 above)
Comment (dmdashenkov) [inline server/src/main/java/io/spine/server/event/EventValidator.java]: Covered now.

**Comment by dmdashenkov** [comment]
@armiol, PTAL again.

**Comment by dmdashenkov** [comment]
@armiol, PTAL again.

File: base/src/main/java/io/spine/util/MessageRejection.java [hunk 13]
```diff
@@ -0,0 +1,53 @@
...
+ * OWNER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL,
+ * SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT
+ * LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE,
+ * DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY
+ * THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+ * (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+ * OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ */
+
+package io.spine.util;
```
Comment (armiol) [inline base/src/main/java/io/spine/util/MessageRejection.java]: The package is still wrong. 

Now. As long as it's related to commands, events and actor request, it should belong to `core` module and `io.spine.core` package.

File: base/src/main/java/io/spine/validate/MessageInvalid.java [hunk 14]
```diff
@@ -0,0 +1,35 @@
...
+
+/**
+ * An interface for the {@link Exception} types which are thrown upon an invalid message.
+ *
+ * <p>Except the methods declared in {@link MessageRejection}, this type is a marker interface.
+ *
+ * @author Dmytro Dashenkov
+ */
+@Internal
+public interface MessageInvalid extends MessageRejection {
```
Comment (armiol) [inline base/src/main/java/io/spine/validate/MessageInvalid.java]: Same as above on package and module. It's not `base` anymore, as it's related to `core` definitions.

File: server/src/main/java/io/spine/server/bus/AbstractBusFilter.java [hunk 15]
```diff
@@ -0,0 +1,42 @@
...
+ * @author Dmytro Dashenkov
+ */
+@Internal
+public abstract class AbstractBusFilter<E extends MessageEnvelope<?>> implements BusFilter<E> {
+
+    @SuppressWarnings("NoopMethodInAbstractClass")
+        // Do nothing by default, but allow customization.
+    @Override
+    public void close() throws Exception {
+        // NoOp
```
Comment (armiol) [inline server/src/main/java/io/spine/server/bus/AbstractBusFilter.java]: Please write a sentence here instead.

File: server/src/main/java/io/spine/server/bus/Bus.java [hunk 16]
```diff
@@ -273,6 +337,76 @@ private void doPost(Iterable<E> envelopes, StreamObserver<IsSent> observer) {
...
+         * @see #appendFilter(BusFilter)
+         */
+        public final Deque<BusFilter<E>> getFilters() {
+            return newLinkedList(filters);
+        }
+
+        /**
+         * Creates new instance of {@code Bus} with the set parameters.
+         *
+         * <p>It is recommended to specify explicitly the resulting type of the bus in the return
```
Comment (armiol) [inline server/src/main/java/io/spine/server/bus/Bus.java]: `It is recommended to specify the exact resulting type...`

File: server/src/main/java/io/spine/server/bus/DeadMessageHandler.java [hunk 17]
```diff
@@ -0,0 +1,47 @@
...
+ * OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ */
+
+package io.spine.server.bus;
+
+import io.spine.annotation.Internal;
+import io.spine.core.MessageEnvelope;
+
+/**
+ * An interface defining a function which handles a dead message.
```
Comment (armiol) [inline server/src/main/java/io/spine/server/bus/DeadMessageHandler.java]: `A definition of a handler for a dead message.` As you noticed, I had removed `function`. Please review the similar cases and modify those accordingly.

File: server/src/main/java/io/spine/server/bus/DeadMessageHandler.java [hunk 18]
```diff
@@ -0,0 +1,47 @@
...
+ *
+ * <p>If a message with no target dispatchers found is passed to the bus, it will result in a call
+ * to this function. The function produces {@link MessageUnhandled} instance describing
+ * the dead message. It may also process the given message (e.g. store it into the bus store).
+ *
+ * @author Dmytro Dashenkov
+ */
+@Internal
+public interface DeadMessageHandler<E extends MessageEnvelope<?>> {
```
Comment (armiol) [inline server/src/main/java/io/spine/server/bus/DeadMessageHandler.java]: Also, let's discuss the naming. We have handlers in `bus` world already. And that's not the one (at least, judging on the super-interface, which is absent). 

From the top of my head, that's `DeadMessageTrap`, which rhymes with the ["Wire Trap" pattern definition](http://www.enterpriseintegrationpatterns.com/patterns/messaging/WireTap.html). In which case, its method should be `capture(...)`

File: base/src/main/java/io/spine/util/MessageRejection.java (hunk 13 above)
Comment (armiol) [inline base/src/main/java/io/spine/util/MessageRejection.java]: @dmdashenkov at a second thought, let's put the whole hierarchy to `io.spine.server.bus` package in the `server` module.

**Comment by dmdashenkov** [comment]
@armiol, PTAL again.
 - `DeadMessageHandler` renamed to `DeadMessageTap`;
 - `MessageRejection` moved to `io.spine.core`.

Let's discuss the packaging preferences (`io.spine.core` vs `io.spine.server.bus`).

File: core/src/main/java/io/spine/core/MessageInvalid.java [hunk 19]
```diff
@@ -0,0 +1,34 @@
...
+ * (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+ * OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ */
+
+package io.spine.core;
+
+import io.spine.annotation.Internal;
+
+/**
+ * An interface for the {@link Exception} types which are thrown upon an invalid message.
```
Comment (armiol) [inline core/src/main/java/io/spine/core/MessageInvalid.java]: Please update this description as well. It's not an exception, it's a message rejection which is caused by the failed validation.

**Pull request merged**